File: samlang-interpreter/src/main/kotlin/samlang/interpreter/MidIrInterpreter.kt
package samlang.interpreter

import samlang.ast.common.ModuleReference
import samlang.ast.mir.MidIrCompilationUnit
import samlang.ast.mir.MidIrExpression
import samlang.ast.mir.MidIrFunction
import samlang.ast.mir.MidIrLoweredExpressionVisitor
import samlang.ast.mir.MidIrLoweredStatementVisitor
import samlang.ast.mir.MidIrNameEncoder
import samlang.ast.mir.MidIrOperator
import samlang.ast.mir.MidIrStatement

/**
 * Interpret [compilationUnit] using [entryModule]'s main function as the main function.
 *
 * @return the printed string.
 */
fun interpretCompilationUnit(compilationUnit: MidIrCompilationUnit, entryModule: ModuleReference): String {
    val environment = setupEnvironment(compilationUnit = compilationUnit)
    val mainFunctionName = MidIrNameEncoder.encodeFunctionName(
        moduleReference = entryModule,
        className = "Main",
        functionName = "main"
    )
    val function = environment.functions[mainFunctionName] ?: error(message = "Bad function.")
    interpretFunction(irFunction = function, environment = environment, arguments = emptyList())
    return environment.printed.toString()
}

private fun setupEnvironment(compilationUnit: MidIrCompilationUnit): GlobalEnvironment {
    val functions = compilationUnit.functions.map { it.functionName to it }.toMap()
    val globalVariables = hashMapOf<String, Long>()
    val strings = hashMapOf<Long, String>()
    var heapPointer = 10000L
    compilationUnit.globalVariables.forEach { (variable, _, content) ->
        val location = heapPointer
        globalVariables[variable.name] = location
        strings[location] = content
        heapPointer += 8
    }
    val functionsGlobals = hashMapOf<Long, String>()
    compilationUnit.functions.forEach { function ->
        val location = heapPointer
        val name = function.functionName
        globalVariables[name] = location
        functionsGlobals[location] = name
        heapPointer += 8
    }
    return GlobalEnvironment(
        functions = functions,
        globalVariables = globalVariables,
        strings = strings,
        functionsGlobals = functionsGlobals,
        heap = hashMapOf(),
        heapPointer = heapPointer,
        printed = StringBuilder()
    )
}

private fun interpretFunction(irFunction: MidIrFunction, arguments: List<Long>, environment: GlobalEnvironment): Long {
    require(value = irFunction.argumentTemps.size == arguments.size)
    val stackFrame = StackFrame()
    irFunction.argumentTemps.map { it.id }.zip(other = arguments).forEach { (id, value) ->
        stackFrame[id] = value
    }
    val statements = irFunction.mainBodyStatements
    val interpreter = MidIrStatementInterpreter(
        environment = environment,
        expressionVisitor = MidIrExpressionInterpreter(environment = environment),
        statements = statements
    )
    var returnedValue: Long? = null
    while (returnedValue == null) {
        val statement = statements[interpreter.programCounter]
        statement.accept(visitor = interpreter, context = stackFrame)
        returnedValue = stackFrame.returnValue
    }
    return returnedValue
}

private class MidIrStatementInterpreter(
    private val environment: GlobalEnvironment,
    private val expressionVisitor: MidIrExpressionInterpreter,
    statements: List<MidIrStatement>
) : MidIrLoweredStatementVisitor<StackFrame, Unit> {
    private var _programCounter: Int = 0
    private val labelMapping: Map<String, Int>

    val programCounter: Int get() = _programCounter

    init {
        labelMapping = hashMapOf<String, Int>().apply {
            statements.forEachIndexed { index, statement ->
                if (statement is MidIrStatement.Label) {
                    this[statement.name] = index
                }
            }
        }
    }

    override fun visit(node: MidIrStatement.MoveTemp, context: StackFrame) {
        context[node.tempId] = node.source.accept(visitor = expressionVisitor, context = context)
        _programCounter++
    }

    override fun visit(node: MidIrStatement.MoveMem, context: StackFrame) {
        environment.heap[node.memLocation.accept(visitor = expressionVisitor, context = context)] =
            node.source.accept(visitor = expressionVisitor, context = context)
        _programCounter++
    }

    override fun visit(node: MidIrStatement.CallFunction, context: StackFrame) {
        val arguments = node.arguments.map { it.accept(visitor = expressionVisitor, context = context) }
        val functionExpression = node.functionExpr
        val functionName = if (functionExpression is MidIrExpression.Name) {
            val functionName = functionExpression.name
            val result = when (functionName) {
                "builtin_malloc" -> {
                    require(value = arguments.size == 1)
                    val size = arguments[0]
                    val start = environment.heapPointer
                    environment.heapPointer += size
                    start
                }
                "builtin_throw" -> {
                    require(value = arguments.size == 1)
                    val argument = arguments[0]
                    val string = environment.strings[argument] ?: error(message = "Bad string at $argument")
                    throw PanicException(reason = string)
                }
                "builtin_stringToInt" -> {
                    require(value = arguments.size == 1)
                    val argument = arguments[0]
                    val string = environment.strings[argument] ?: error(message = "Bad string at $argument")
                    string.toLongOrNull() ?: throw PanicException(reason = "Bad string: $string")
                }
                "builtin_intToString" -> {
                    require(value = arguments.size == 1)
                    val stringForm = arguments[0].toString()
                    val location = environment.heapPointer
                    environment.heapPointer += 8
                    environment.strings[location] = stringForm
                    location
                }
                "builtin_println" -> {
                    require(value = arguments.size == 1)
                    val argument = arguments[0]
                    val string = environment.strings[argument] ?: error(message = "Bad string at $argument")
                    environment.printed.append(string).append('\n')
                    0L
                }
                else -> null
            }
            if (result != null) {
                node.returnCollector?.let { context[it.id] = result }
                _programCounter++
                return
            }
            functionName
        } else {
            val functionAddress = functionExpression.accept(visitor = expressionVisitor, context = context)
            environment.functionsGlobals[functionAddress]
                ?: error(message = "Undefined function at $functionAddress! Expression: $functionExpression")
        }
        val function = environment.functions[functionName]
            ?: error(message = "Missing function $functionName")
        val result = interpretFunction(
            irFunction = function,
            arguments = arguments,
            environment = environment
        )
        node.returnCollector?.let { context[it.id] = result }
        _programCounter++
    }

    override fun visit(node: MidIrStatement.Jump, context: StackFrame) {
        val label = node.label
        _programCounter = (labelMapping[label] ?: error("BAD label: $label!"))
    }

    override fun visit(node: MidIrStatement.ConditionalJumpFallThrough, context: StackFrame) {
        if (node.condition.accept(visitor = expressionVisitor, context = context) != 0L) {
            visit(node = MidIrStatement.Jump(label = node.label1), context = context)
        } else {
            _programCounter++
        }
    }

    override fun visit(node: MidIrStatement.Label, context: StackFrame) {
        _programCounter++
    }

    override fun visit(node: MidIrStatement.Return, context: StackFrame) {
        val value = node.returnedExpression?.accept(visitor = expressionVisitor, context = context) ?: 0L
        context.returnValue(value = value)
    }
}

private class MidIrExpressionInterpreter(
    private val environment: GlobalEnvironment
) : MidIrLoweredExpressionVisitor<StackFrame, Long> {
    override fun visit(node: MidIrExpression.Constant, context: StackFrame): Long = node.value

    override fun visit(node: MidIrExpression.Name, context: StackFrame): Long =
        environment.globalVariables[node.name] ?: error(message = "Referencing undefined global ${node.name}!")

    override fun visit(node: MidIrExpression.Temporary, context: StackFrame): Long = context[node.id]

    override fun visit(node: MidIrExpression.Op, context: StackFrame): Long {
        val value1 = node.e1.accept(visitor = this, context = context)
        val value2 = node.e2.accept(visitor = this, context = context)
        return when (node.operator) {
            MidIrOperator.MUL -> value1 * value2
            MidIrOperator.DIV -> {
                if (value2 == 0L) {
                    throw PanicException(reason = "Division by zero!")
                }
                value1 / value2
            }
            MidIrOperator.MOD -> {
                if (value2 == 0L) {
                    throw PanicException(reason = "Division by zero!")
                }
                value1 % value2
            }
            MidIrOperator.ADD -> value1 + value2
            MidIrOperator.SUB -> value1 - value2
            MidIrOperator.XOR -> value1 xor value2
            MidIrOperator.LT -> toInt(boolean = value1 < value2)
            MidIrOperator.LE -> toInt(boolean = value1 <= value2)
            MidIrOperator.GT -> toInt(boolean = value1 > value2)
            MidIrOperator.GE -> toInt(boolean = value1 >= value2)
            MidIrOperator.EQ -> toInt(boolean = value1 == value2)
            MidIrOperator.NE -> toInt(boolean = value1 != value2)
            MidIrOperator.AND -> if (value1 == 0L) 0L else if (value2 == 0L) 0L else 1L
            MidIrOperator.OR -> if (value1 != 0L) 1L else if (value2 != 0L) 1L else 0L
        }
    }

    private fun toInt(boolean: Boolean): Long = if (boolean) 1L else 0L

    override fun visit(node: MidIrExpression.Mem, context: StackFrame): Long =
        environment.heap[node.expression.accept(visitor = this, context = context)] ?: 0
}

private class GlobalEnvironment(
    val functions: Map<String, MidIrFunction>,
    val globalVariables: Map<String, Long>,
    val functionsGlobals: Map<Long, String>,
    val strings: MutableMap<Long, String>,
    val heap: MutableMap<Long, Long>,
    var heapPointer: Long,
    val printed: StringBuilder
)

private class StackFrame {
    private val variables: MutableMap<String, Long> = hashMapOf()
    private var _returnedValue: Long? = null

    val returnValue: Long? get() = _returnedValue

    operator fun get(name: String): Long = variables[name] ?: 0

    operator fun set(name: String, value: Long) {
        variables[name] = value
    }

    fun returnValue(value: Long) {
        _returnedValue = value
    }
}

File: samlang-core/interpreter/mid-ir-interpreter.ts
/* eslint-disable no-param-reassign */

import {
  ENCODED_FUNCTION_NAME_MALLOC,
  ENCODED_FUNCTION_NAME_THROW,
  ENCODED_FUNCTION_NAME_STRING_TO_INT,
  ENCODED_FUNCTION_NAME_INT_TO_STRING,
  ENCODED_FUNCTION_NAME_STRING_CONCAT,
  ENCODED_FUNCTION_NAME_PRINTLN,
  ENCODED_COMPILED_PROGRAM_MAIN,
} from '../ast/common-names';
import type { MidIRExpression, MidIRFunction, MidIRCompilationUnit } from '../ast/mir-nodes';
import PanicException from './panic-exception';

class StackFrame {
  private variables = new Map<string, bigint>();

  private _returnValue: bigint | null = null;

  get returnValue(): bigint | null {
    return this._returnValue;
  }

  setReturnValue(value: bigint): void {
    this._returnValue = value;
  }

  getLocalValue(name: string): bigint {
    return this.variables.get(name) ?? BigInt(0);
  }

  setLocalValue(name: string, value: bigint) {
    this.variables.set(name, value);
  }
}

type MidIRInterpreterMutableGlobalEnvironment = {
  // A collection of all available functions.
  readonly functions: ReadonlyMap<string, MidIRFunction>;
  // Global variable name to fake address mapping.
  readonly globalVariables: ReadonlyMap<string, bigint>;
  // Fake function address to function name mapping.
  readonly functionsGlobals: ReadonlyMap<bigint, string>;
  // Strings generated at compile time and runtime.
  readonly strings: Map<bigint, string>;
  // Address to value mapping of heap.
  readonly heap: Map<bigint, bigint>;
  heapPointer: bigint;
  // A collection of already printed stuff.
  printed: string;
};

const interpretMidIRExpression = (
  environment: MidIRInterpreterMutableGlobalEnvironment,
  stackFrame: StackFrame,
  expression: MidIRExpression
): bigint => {
  switch (expression.__type__) {
    case 'MidIRConstantExpression':
      return expression.value;
    case 'MidIRNameExpression': {
      const value = environment.globalVariables.get(expression.name);
      // istanbul ignore next
      if (value == null) throw new Error(`Referencing undefined global ${expression.name}.`);
      return value;
    }
    case 'MidIRTemporaryExpression':
      return stackFrame.getLocalValue(expression.temporaryID);
    case 'MidIRImmutableMemoryExpression': {
      const value = environment.heap.get(
        interpretMidIRExpression(environment, stackFrame, expression.indexExpression)
      );
      // istanbul ignore next
      if (value == null) throw new Error();
      return value;
    }
    case 'MidIRBinaryExpression': {
      const value1 = interpretMidIRExpression(environment, stackFrame, expression.e1);
      const value2 = interpretMidIRExpression(environment, stackFrame, expression.e2);
      switch (expression.operator) {
        case '+':
          return value1 + value2;
        case '-':
          return value1 - value2;
        case '*':
          return value1 * value2;
        case '/':
          if (value2 === BigInt(0)) throw new PanicException('Division by zero!');
          return value1 / value2;
        case '%':
          if (value2 === BigInt(0)) throw new PanicException('Mod by zero!');
          return value1 % value2;
        case '^':
          // eslint-disable-next-line no-bitwise
          return value1 ^ value2;
        case '<':
          return BigInt(value1 < value2);
        case '<=':
          return BigInt(value1 <= value2);
        case '>':
          return BigInt(value1 > value2);
        case '>=':
          return BigInt(value1 >= value2);
        case '==':
          return BigInt(value1 === value2);
        case '!=':
          return BigInt(value1 !== value2);
      }
    }
  }
};

const interpretMidIRFunction = (
  environment: MidIRInterpreterMutableGlobalEnvironment,
  midIRFunction: MidIRFunction,
  functionArguments: readonly bigint[]
): bigint => {
  // istanbul ignore next
  if (functionArguments.length !== midIRFunction.argumentNames.length) throw new Error();
  const stackFrame = new StackFrame();
  midIRFunction.argumentNames.forEach((argumentName, index) => {
    stackFrame.setLocalValue(argumentName, functionArguments[index]);
  });

  let programCounter = 0;
  const labelMapping = new Map<string, number>();
  midIRFunction.mainBodyStatements.forEach((statement, index) => {
    if (statement.__type__ === 'MidIRLabelStatement') {
      labelMapping.set(statement.name, index);
    }
  });

  let returnedValue: bigint | null = null;
  while (returnedValue == null) {
    const statementToInterpret = midIRFunction.mainBodyStatements[programCounter];

    switch (statementToInterpret.__type__) {
      case 'MidIRMoveTempStatement':
        stackFrame.setLocalValue(
          statementToInterpret.temporaryID,
          interpretMidIRExpression(environment, stackFrame, statementToInterpret.source)
        );
        programCounter += 1;
        break;

      case 'MidIRMoveMemStatement':
        environment.heap.set(
          interpretMidIRExpression(
            environment,
            stackFrame,
            statementToInterpret.memoryIndexExpression
          ),
          interpretMidIRExpression(environment, stackFrame, statementToInterpret.source)
        );
        programCounter += 1;
        break;

      case 'MidIRJumpStatement': {
        const target = labelMapping.get(statementToInterpret.label);
        // istanbul ignore next
        if (target == null) throw new Error(`Bad label ${statementToInterpret.label}!`);
        programCounter = target;
        break;
      }

      case 'MidIRLabelStatement':
        programCounter += 1;
        break;

      case 'MidIRConditionalJumpFallThrough':
        if (
          interpretMidIRExpression(
            environment,
            stackFrame,
            statementToInterpret.conditionExpression
          ) !== BigInt(0)
        ) {
          const target = labelMapping.get(statementToInterpret.label1);
          // istanbul ignore next
          if (target == null) throw new Error(`Bad label ${statementToInterpret.label1}!`);
          programCounter = target;
        } else {
          programCounter += 1;
        }
        break;

      case 'MidIRReturnStatement':
        stackFrame.setReturnValue(
          statementToInterpret.returnedExpression != null
            ? interpretMidIRExpression(
                environment,
                stackFrame,
                statementToInterpret.returnedExpression
              )
            : BigInt(0)
        );
        break;

      case 'MidIRCallFunctionStatement': {
        const functionArgumentValues = statementToInterpret.functionArguments.map((it) =>
          interpretMidIRExpression(environment, stackFrame, it)
        );
        const functionExpression = statementToInterpret.functionExpression;
        let functionName: string;
        if (functionExpression.__type__ === 'MidIRNameExpression') {
          functionName = functionExpression.name;
          let result: bigint | null;
          switch (functionName) {
            case ENCODED_FUNCTION_NAME_MALLOC: {
              const start = environment.heapPointer;
              environment.heapPointer += functionArgumentValues[0];
              result = start;
              break;
            }
            case ENCODED_FUNCTION_NAME_THROW: {
              const string = environment.strings.get(functionArgumentValues[0]);
              // istanbul ignore next
              if (string == null) throw new Error('Bad string!');
              throw new PanicException(string);
            }
            case ENCODED_FUNCTION_NAME_STRING_TO_INT: {
              const string = environment.strings.get(functionArgumentValues[0]);
              // istanbul ignore next
              if (string == null) throw new Error('Bad string!');
              try {
                result = BigInt(string);
                break;
              } catch {
                throw new PanicException(`Bad string: ${string}`);
              }
            }
            case ENCODED_FUNCTION_NAME_INT_TO_STRING: {
              const string = String(functionArgumentValues[0]);
              const location = environment.heapPointer;
              environment.heapPointer += BigInt(8);
              environment.strings.set(location, string);
              result = location;
              break;
            }
            case ENCODED_FUNCTION_NAME_STRING_CONCAT: {
              const string1 = environment.strings.get(functionArgumentValues[0]);
              const string2 = environment.strings.get(functionArgumentValues[1]);
              // istanbul ignore next
              if (string1 == null || string2 == null) throw new Error('Bad string');
              const location = environment.heapPointer;
              environment.heapPointer += BigInt(8);
              environment.strings.set(location, string1 + string2);
              result = location;
              break;
            }
            case ENCODED_FUNCTION_NAME_PRINTLN: {
              const string = environment.strings.get(functionArgumentValues[0]);
              // istanbul ignore next
              if (string == null) throw new Error('Bad string!');
              environment.printed += `${string}\n`;
              result = BigInt(0);
              break;
            }
            default:
              result = null;
          }
          if (result != null) {
            if (statementToInterpret.returnCollectorTemporaryID != null) {
              stackFrame.setLocalValue(statementToInterpret.returnCollectorTemporaryID, result);
            }
            programCounter += 1;
            break;
          }
        } else {
          const functionAddress = interpretMidIRExpression(
            environment,
            stackFrame,
            functionExpression
          );
          const nullableName = environment.functionsGlobals.get(functionAddress);
          // istanbul ignore next
          if (nullableName == null) throw new Error(`Undefined function at ${functionAddress}!`);
          functionName = nullableName;
        }

        const functionToCall = environment.functions.get(functionName);
        // istanbul ignore next
        if (functionToCall == null) throw new Error(`Missing function ${functionName}`);
        const result = interpretMidIRFunction(environment, functionToCall, functionArgumentValues);
        if (statementToInterpret.returnCollectorTemporaryID != null) {
          stackFrame.setLocalValue(statementToInterpret.returnCollectorTemporaryID, result);
        }
        programCounter += 1;
        break;
      }
    }

    returnedValue = stackFrame.returnValue;
  }

  return returnedValue;
};

const setupMidIRCompilationUnitIntepretationEnvironment = (
  compilationUnit: MidIRCompilationUnit
): MidIRInterpreterMutableGlobalEnvironment => {
  const functions = new Map(compilationUnit.functions.map((it) => [it.functionName, it]));
  const globalVariables = new Map<string, bigint>();
  const strings = new Map<bigint, string>();
  let heapPointer = BigInt(10000);
  compilationUnit.globalVariables.forEach(({ name, content }) => {
    const location = heapPointer;
    globalVariables.set(name, location);
    strings.set(location + BigInt(8), content);
    heapPointer += BigInt(8);
  });
  const functionsGlobals = new Map<bigint, string>();
  compilationUnit.functions.forEach(({ functionName }) => {
    const location = heapPointer;
    globalVariables.set(functionName, location);
    functionsGlobals.set(location, functionName);
    heapPointer += BigInt(8);
  });
  return {
    functions,
    globalVariables,
    strings,
    functionsGlobals,
    heap: new Map(),
    heapPointer,
    printed: '',
  };
};

const interpretMidIRCompilationUnit = (compilationUnit: MidIRCompilationUnit): string => {
  const environment = setupMidIRCompilationUnitIntepretationEnvironment(compilationUnit);
  const mainFunction = environment.functions.get(ENCODED_COMPILED_PROGRAM_MAIN);
  // istanbul ignore next
  if (mainFunction == null) throw new Error('Missing new function!');
  interpretMidIRFunction(environment, mainFunction, []);
  return environment.printed;
};

export default interpretMidIRCompilationUnit;
